enterprise/backend/src/metabase_enterprise/caching/strategies.clj

```diff
@@ -0,0 +1,218 @@
+  [card dashboard database]
+  (let [ttls              [(:cache_ttl card) (:cache_ttl dashboard) (:cache_ttl database)]
+        most-granular-ttl (first (filter some? ttls))]
+    (when most-granular-ttl ; stored TTLs are in hours; convert to seconds
+      (* most-granular-ttl 3600))))
+
+(defenterprise-schema granular-cache-strategy :- CacheStrategy
+  "Returns the granular cache strategy for a card."
+  :feature :cache-granular-controls
+  [card dashboard-id]
+  (let [qs     [{:from   [:cache_config]
+                 :select [:id]
+                 :where  [:and [:= :model "question"]   [:= :model_id (:id card)]]}
+                {:from   [:cache_config]
+                 :select [:id]
+                 :where  [:and [:= :model "dashboard"]  [:= :model_id dashboard-id]]}
+                {:from   [:cache_config]
+                 :select [:id]
+                 :where  [:and [:= :model "collection"] [:= :model_id (:collection_id card)]]}
+                {:from   [:cache_config]
+                 :select [:id]
+                 :where  [:and [:= :model "database"]   [:= :model_id (:database_id card)]]}
+                {:from   [:cache_config]
+                 :select [:id]
+                 :where  [:= :model "root"]}]
+        q      {:select [:id]
+                :limit  1
+                :from   [[{:union-all qs} :unused_alias]]}
+        config (t2/select-one :model/CacheConfig :id q)]
+    (when (public-settings/enable-query-caching)
+      (if config
+        (merge {:type       (:strategy config)
+                :updated_at (:updated_at config)
+                :payload    (:payload config)}
+               (:config config))
+
+        (if-let [ttl (granular-ttl
+                      card
+                      (t2/select-one [:model/Dashboard :cache_ttl] :id dashboard-id)
+                      (t2/select-one [:model/Database :cache_ttl] :id (:database_id card)))]
+          {:type     :duration
+           :duration (int (/ ttl 3600)) ;; UI accepts hours, so it should divide cleanly
+           :unit     "hours"}
+          {:type         :ttl
+           :multiplier   (public-settings/query-caching-ttl-ratio)
+           :min_duration 50})))))
+
+;;; Strategy execution
+
+(defmulti fetch-cache-stmt-ee*
+  "Generate prepared statement for a db cache backend for a given strategy"
+  (fn [strategy _hash _conn] (:type strategy)))
+
+(defmethod fetch-cache-stmt-ee* :ttl [strategy query-hash conn]
+  (backend.db/fetch-cache-stmt-ttl strategy query-hash conn))
+
+(defmethod fetch-cache-stmt-ee* :duration [strategy query-hash conn]
+  (assert (:duration strategy))
+  (assert (:unit strategy))
+  (let [duration  (t/duration (:duration strategy) (keyword (:unit strategy)))
+        timestamp (t/minus (t/offset-date-time) duration)]
+    (backend.db/prepare-statement conn query-hash timestamp)))
+
+;; NOTE: something changes `:updated_at` to `:updated-at`
```

Comment:
Not really, it's something inside of query processor. I thought that it's `qp.middleware.normalize-query`, but nope, it receives kebab-case already. I did not dig really deep since there is no way to disable that right now, it just was surprising. :)